### ISSUE TITLE
Implement MVP testcase for handsontable extension

### ISSUE THREAD

**Issue by benib** [nzzdev/Q-editor#111]
### Description

We want to know how we can develop a custom extension for handsontable to be used in schema-editor-table

The minimum viable product is a `div` element opening after right-click into a cell in the table editor.

**Comment by philipkueng** [comment]
@manuelroth and I disgusted about the way we could implement this feature yesterday. We came to the conclusion that we have different approaches to save the data and want to know what you think about it @benib 

### Values and meta-data in item
In this case we would alter the schema where the items-array will contain an object, not just a string. The object would contain a value-property and a meta-data property, containing which labeling the value has.

#### Pro
- Value and meta-data would be stored together 
- No need to map things when loading/saving

#### Contra
- schema of all Q-Elements containing data-table would need some changes

**Comment by benib** [comment]
I would go with another variante of `Values and meta-data in item`

schema.json would look like this:
```
"data": {
  "title": "Daten",
  "type": "array",
  "Q:type": "table",
  "Q:options": {
    "metaData": {
      "property": "somePropertyName",
      "cells": {
        "schema": {
          "type": "object",
          "properties": {}
            "highlight": {
              "type": "boolean"
            },
            "annotation": {
              "type": "string"
            }
          }
        } 
        "notOnColumn": 1,
        "notOnRow": 1
      },
      "rows": false,
      "columns": {
        "notOnColumn": 1,
        "schema": {
          "type": "object",
          "properties": {}
            "highlight": {
              "type": "boolean"
            }
          }
        } 
      },
    }
  }
}
```
where `Q:options.metadata.cells`, `Q:options.metadata.rows` and `Q:options.metadata.columns` could be `false` or `undefined` to not allow any metadata or an `object` with more information where to allow/disallow this and a `schema` describing the the metaData.

`schema-editor-table` would then store the metaData on the item property defined in `Q:options.metadata.property`. In this case `somePropertyName`. This would result in item data like this:

```
{
  "data": [
    [
      "Day",
      "A"
    ],
    [
      "01.01.2018",
      "499"
    ],
    [
      "02.01.2018",
      "250"
    ]
  ],
  "somePropertyName": {
    cells: [
      [null, null],
      [
        null,
        {
          highlight: true,
          annotation: "some string"
        }
      ]
    ],
    columns: [
      null,
      {
        highlight: true
      }
    ]
  }
}
```

what do you think?

**Comment by benib** [comment]
I am not yet sure if I prefer the shadow-array storing format as described in the comment above or one that only stores the actual metadata with the coordinates. if we go with the coordinate system I'd suggest a slight change of the format from your suggestion @philipkueng to not have to transform the data coming from `schema-editor` with the given `metaData.cells.schema`.

this would then look something like this
```
{
  "data": [
    [
      "Day",
      "A"
    ],
    [
      "01.01.2018",
      "499"
    ],
    [
      "02.01.2018",
      "250"
    ]
  ],
  "somePropertyName": {
    cells: [
      {
        rowIndex: 1,
        colIndex: 1,
        data: {
          highlight: true,
          annotation: "some string"
        }
      }
    ],
    columns: [
      {
        colIndex: 1,
        data: {
          highlight: true
        }
      }
    ]
  }
}
```

Regarding the handling of the data: it's imho more simple to have a shadow-array. That way any metaData is right at hand without any looping over an array while looping through the `data` array.

Regarding the amount of data to be stored and sent from Q-editor to Q-server: I am not completely convinced that this is a problem. Even with 1000 rows and 3 columns this is not too much data, and it has never to be sent to readers clients.

I will think about this some more. Please do so as well.

**Comment by manuelroth** [comment]
I like the idea of only storing the metadata with the coordinates for cells/rows/columns which are actually annotated/highlighted. I think it doesn't add a huge overhead to make sure that there are not multiple metadata entries for each type.

I think if we want to allow different metadata based on the type (cell/column/row) we should go for Variant B.

**Comment by benib** [comment]
After thinking about this some more:

As we do not have the complete item in schema-editor-table but just to property `a` that is defined as a table type, we cannot easily store the metaData in whatever property we like on the item without passing the complete item into schema-editor-table.

this was built like this to make it impossible to have side effects on the item inside of one schema-editor type.

I'd like to keep it that way.

this has the consequence that we need to store the metaData somewhere on the property `a` together with the data.

for example:

```
a: {
  table: [[],[],[]],
  meta: [
  ]
}
```

- The format in `a.metaData` should probably follow the Variant B explained above.
- We can remove the `Q:options.metaData.property` property and just hardcode them to table and meta in my opinion. I am willing to discuss the naming here.

This means that introducing `Q:options.metaData` on an existing tool using schema-editor-table is a breaking change and needs a migration. See https://nzzdev.github.io/Q-server/migrations.html about how to write the migration in the tool and how to run it.

We lack a system to automatically migrate the data of an item in case the schema doesn't match, so we should run this on prod in some low-profile time and with good preparation and orchestration.

/ @philipkueng @manuelroth

**Comment by manuelroth** [comment]
- Just to be sure. You are proposing that the type of the [data](https://github.com/nzzdev/Q-table/blob/dev/resources/schema.json#L16) property should be changed from `array` to `object`. And  has two properties (table and meta)?
- What do you think about keeping the `data` property type array and having the first item be the actual data (table) and second item be the metadata (meta). This has the benefit of being very easily extensible in the future as the schema doesn't need to be altered if more things get added to the array. The migration is also needed with this approach.

Example item: 
```json
{
  "data": [[["Header1", "Header2"], [1,2,3], [4,5,6]], {"cells": [], "columns": [], "rows": []}]
}
```

Schema:
```json
"data": {
      "title": "Daten",
      "type": "array",
      "Q:type": "table",
      "Q:options": {
        "allowTranspose": false
      },
      "items": {
        "type": "array",
        "items": {
          "anyOf": [
            {
            	"type": "array"
          	},
           {
            	"type": "object"
          	}
          ]
        }
      },
      "minItems": 2
    }

**Comment by manuelroth** [comment]
Yes I agree. Its more explicit and altering the schema is not a big deal. Lets go with the object. I think it makes sense to close this issue and open a new issue describing the data structure we have decided on. @philipkueng could you do that?
